units/generate_tasks.go

```diff
@@ -50,75 +46,99 @@ func makeGenerateTaskJob() *generateTasksJob {
+	if err != nil && adb.ResultsNotFound(err) {
+		return err
+	}
+	if err != nil {
+		return errors.Wrap(err, "error saving config in `generate.tasks`")
+	}
+	return nil
+}
+
 func (j *generateTasksJob) Run(ctx context.Context) {
 	defer j.MarkComplete()
 	start := time.Now()
 
-	projects, err := parseProjects(j.JSON)
+	t, err := task.FindOneId(j.TaskID)
 	if err != nil {
-		j.AddError(errors.Wrap(err, "error parsing JSON from `generate.tasks`"))
-		return
+		j.AddError(errors.Wrapf(err, "problem finding task %s", j.TaskID))
+	}
+	if t == nil {
+		j.AddError(errors.Errorf("task %s does not exist", j.TaskID))
+	}
+
+	err = j.generate(ctx, t)
+	if err != nil && !adb.ResultsNotFound(err) {
+		j.AddError(err)
+		j.AddError(t.SetGenerateTasksError(err))
+	}
+	if err == nil {
+		j.AddError(task.MarkGeneratedTasks(j.TaskID))
 	}
-	g := model.MergeGeneratedProjects(projects)
-	g.TaskID = j.TaskID
 
-	var attempt int
-	err = util.Retry(
-		ctx,
-		func() (bool, error) {
-			attemptStart := time.Now()
-			attempt++
-
-			t, err := task.FindOneId(j.TaskID)
-			if err != nil {
-				return false, err
-			}
-			if t == nil {
-				return false, errors.Errorf("unable to find task %s", g.TaskID)
-			}
-			if t.GeneratedTasks {
-				return false, nil // already generated tasks, noop
-			}
-			p, v, t, pm, err := g.NewVersion() // nolint
-			if err != nil {
-				return false, err
-			}
-			if err = validator.CheckProjectConfigurationIsValid(p); err != nil {
-				return false, err
-			}
-			err = g.Save(ctx, p, v, t, pm)
-			if err != nil && adb.ResultsNotFound(err) {
-				return true, gimlet.ErrorResponse{
-					StatusCode: http.StatusInternalServerError,
-					Message:    errors.Wrap(err, "error updating config in `generate.tasks`").Error(),
-				}
-			}
-			if err = t.MarkGeneratedTasks(); err != nil {
-				return true, gimlet.ErrorResponse{
-					StatusCode: http.StatusInternalServerError,
-					Message:    errors.Wrapf(err, "problem marking task '%s' as having generated tasks", t.Id).Error(),
-				}
-			}
-			grip.Info(message.Fields{
-				"message":               "generate.tasks succeeded",
-				"attempt":               attempt,
-				"attempt_duration_secs": time.Since(attemptStart).Seconds(),
-				"total_duration_secs":   time.Since(start).Seconds(),
-				"task":                  t.Id,
-				"version":               t.Version,
-			})
-			return false, nil
-		}, 100, time.Second, 15*time.Second)
-	j.AddError(err)
+	grip.InfoWhen(err == nil, message.Fields{
+		"message":       "generate.tasks finished",
+		"duration_secs": time.Since(start).Seconds(),
+		"task":          t.Id,
+		"version":       t.Version,
+	})
+	grip.DebugWhen(adb.ResultsNotFound(err), message.Fields{
+		"message":       "generate.tasks noop",
+		"duration_secs": time.Since(start).Seconds(),
+		"task":          t.Id,
+		"version":       t.Version,
+	})
+	grip.ErrorWhen(err != nil, message.WrapError(err, message.Fields{
```

Comment:
if the above case (debug) runs, then this will also run. 

also message.WrapError() is a noop when the err is nil, so your when clause is redundant.